File: 3876-construct-uniform-parity-array-ii/3876-construct-uniform-parity-array-ii.java
class Solution {
    public boolean uniformArray(int[] nums1) {
        int smallestOdd = Integer.MAX_VALUE;
        int smallestEven = Integer.MAX_VALUE;

        for (int num : nums1) {
            if ((num & 1) == 0)
                smallestEven = Math.min(smallestEven, num);
            else
                smallestOdd = Math.min(smallestOdd, num);
        }

        // all elements already have the same parity
        if (smallestOdd == Integer.MAX_VALUE || smallestEven == Integer.MAX_VALUE)
            return true;
        
        // mixed parity: make everything odd
        return smallestOdd < smallestEven;
    }
}

File: 3876-construct-uniform-parity-array-ii/3876-construct-uniform-parity-array-ii.py
class Solution:
    def uniformArray(self, nums1: list[int]) -> bool:
        smallestOdd = float('inf')
        smallestEven = float('inf')

        for num in nums1:
            if (num & 1) == 0:
                smallestEven = min(smallestEven, num)
            else:
                smallestOdd = min(smallestOdd, num)

        # all elements already have the same parity
        if smallestOdd == float('inf') or smallestEven == float('inf'):
            return True
        
        # mixed parity: make everything odd
        return smallestOdd < smallestEven
